backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConnectionFactory.java

```diff
@@ -18,30 +18,47 @@
  ****************************************************************/
 package org.apache.james.backend.rabbitmq;
 
-import java.util.concurrent.ExecutionException;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.Callable;
 
 import javax.inject.Inject;
 
-import org.apache.james.util.retry.RetryExecutorUtil;
-
-import com.nurkiewicz.asyncretry.AsyncRetryExecutor;
 import com.rabbitmq.client.Connection;
 import com.rabbitmq.client.ConnectionFactory;
 
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
 public class RabbitMQConnectionFactory {
+    private class ConnectionCallable implements Callable<Connection> {
+        private final Callable<Connection> wrapped;
+        private Optional<Connection> connection;
+
+        ConnectionCallable(Callable<Connection> wrapped) {
+            this.wrapped = wrapped;
+            connection = Optional.empty();
+        }
+
+        @Override
+        public Connection call() throws Exception {
+            if (connection.map(Connection::isOpen).orElse(false)) {
+                return connection.get();
+            }
+            Connection newConnection = wrapped.call();
```

Comment:
`wrapped` looks more like a supplier than a Callable. Dont you think? And I suggest to name it to `connectionSupplier`

Reply:
A callable is a supplier that can throw exception...

Let's review the naming later on.

Here we do have an issue with key registration as well.... Temporary queues have the good idea to disapear upon connection failures...

Reply:
I think making them to `non auto delete` and deleting queue manually when unregistering key registration could be a choice.

Reply:
> Temporary queues have the good idea to disapear upon connection failures...

can you add a test for key dispatching to describe this issue?

Reply:
Added